### DISCUSSION TITLE
Auto freeze dependencies to track for Received calls

### DISCUSSION THREAD

**Discussion by dmtrrk** [AutoFixture/AutoFixture#1321] [Q&A]
Not sure whether it is a bug or by design.
I have a simple code

```
interface IInterface {
  void Do();
}

class MyTest {
  private IInterface _dep;
  public MyTest(IInterface dep) { _dep = dep; }

  public void Test() {
    _dep.Do();
  }
}

// Test
var sut = fixture.Create<MyTest>();

sut.Test();

fixture.Freeze<IInterface>().Received(1).Do();
```

This code does not work, freezing should happen before the call. What is the logic behind this?
To use this approach, I'll have to freeze it

```
var freeze = fixture.Freeze<IInterface>();
//...
sut.Test();

freeze.Received(1).Do();
```

Is there a customization that would freeze automatically all class dependencies on resolution, so I don't need to freeze manually?

**Comment by aivascu** [comment]
@deeptowncitizen this is by design. 

When a type frozen, the same instance will be resolved for any class that has a dependency on it. This is similar to registering a singleton in an IoC container. Frozen types are intended to resolve to the same instance only **after** they've been frozen.

To answer your second question there is a customization that could give you the behavior you want, but it feels like I would be giving you bad advice, if I would tell you. Could you first explain why would you expect all instances of every type to be frozen?

BTW if unsure about some behavior in the library please open a question in the discussions, and to do not ignore the templates when filing issues.

**Reply by dmtrrk** [reply]
thank you for the advice. tbh I did not know about the Discussion tab.

Let's assume, that I want to perform a unit test, and I want to test a single class. In my scope, I want to isolate my class from any external dependencies to focus only on its behavior. Then, let's assume that my class does not instantiate any other non-data objects, but is using DI to inject dependencies.
Once I tell that all tests in my scope will have such characteristics, I consider that there is no need to resolve the same interface from the class dependencies twice, and it can be frozen by default.
In addition to the default logic that mocks all dependencies, I have additional customization that `SubstituteRelay` all concrete dependencies with virtual methods. And I expect to have them frozen by default as well, so the behavior of mocked methods will be the same.

**Reply by aivascu** [reply]
@deeptowncitizen I think I understand what you're saying, though it is very unusual.

1. Are you always interested in all mocks? You are usually interested enough in a mock object, to freeze it, when you either want to setup the behavior or when you want to inspect its state after some actions. Having all mocks frozen would mean that you are always using all the mocks in your tests, which in my experience is very rare.
2. Are you never using interfaces for anything other than contracts to be injected in these "services"? Will you ever encounter a collection or sequence of objects of distinct types but which all implement the same interface. Because if you do then all the elements in that sequence will be generated by AutoFixture as the same object.
```cs
public MyObservale(IEnumerable<IObserver>  observers) { /*...*/ }
```
3. Are you using AutoFixture purely as an auto-mocking container that will resolve all injected instances as mocks? Meaning you never make use of AutoFixture's data generation capabilities? If that is the case then there are simpler and more performant ways to achieve that using a conventional IoC container, like AutoFac or SimpleInject. AutoFixture is a more complex tool that allows to also generate data, and if you're not using that feature, then using it at all is a waste of resources.

**Reply by dmtrrk** [reply]
1. 

> Are you always interested in all mocks?

In most scenarios yes:
- all ctor dependencies to be mocked by default
- all 2-level objects (object created by the factory that was injected to the ctor)

At the same time I see there can be some scenarios that will not fit to this logic:
- multiple data objects
- sequence of concrete objects of the same interface
- having some circular resolution of the same type from different points (rare). It's more like to call the factory multiple times.

But these issues are rare and can be easily solved with `Create+Inject` or `Customize` logic.

> You are usually interested enough in a mock object, to freeze it, when you either want to setup the behavior or when you want to inspect its state after some actions.

That's true. But if I don't need it, the mocked object does not affect anything. maybe the memory, but I'm not sure anyone would care.

> Having all mocks frozen would mean that you are always using all the mocks in your tests, which in my experience is very rare.

From my understanding this means:
- the same type will be resolved with the same object (that can be implicit and lead to unexpected behavior)
- there will be some mocked object to track for some actions, that will never be validated (not sure I care about this overhead)

2. 

> Are you never using interfaces for anything other than contracts to be injected in these "services"?

As I mentioned, this is one of the potential issues. But in my codebase the chance to have such collection is significantly less than a simple abstraction (not sure who has something different, even if the application is very-very reactive).
In any case, as I mentioned this case is identified very easily and can be solved with the `Create+Inject` pair.
Also, I believe we are talking about some customization that would auto-mock some/all types. In this case I believe I can have some smart flag to disable it in the test or exclude some types, etc.

3.

> Are you using AutoFixture purely as an auto-mocking container that will resolve all injected instances as mocks?

Yes, in 80% of cases.

> AutoFixture is a more complex tool that allows to also generate data, and if you're not using that feature, then using it at all is a waste of resources

I'm using this for testing only. I don't see any feasible performance degradation. But I have less fragile unit tests with a very laconic code.. Today I'm not generating a lot of test data, but have plans to use it more. tbh I did not try to implement the same functionality with the regular containers, but AF + <mocking lib> sounds more reasonable for this purpose as it is already written (and thanks all maintainers for that)

**Reply by aivascu** [reply]
@deeptowncitizen by the looks of it you don't really care if the unused mocks are frozen or not. So wouldn't it make more sense to explicitly specify which mocks you want frozen in each test? As mentioned previously this would avoid some potential issues, when the same instance will be resolved twice.

Anyway onto the solution you were seeking. I came up with something, but I really don't recommend your write tests the way you are at the moment. It is better to stick to the recommended API usage. You'll avoid issues, you'll have better designed tests and you'll generally be able to troubleshoot and find support when you have questions.

I originally was going to propose that you use the `FreezeOnMatchCustomization`, since I vaguely remembered that it uses a `IRequestSpecification` to check if it should freeze the request. So in my idea you should have been able to freeze anything by providing a `TrueRequestSpecification` in the constructor. When I looked at the implementation it turned out it requires a request to freeze in the constructor, but you require to match everything.

I tried a few different implementations and concluded that the best way to achieve your required behavior is to create a... "Behavior" 🙂, since behaviors are the only way to wrap the entire fixture pipeline and allow only one pass-through for a request. This is not an out-of-the-box solution, but it will do what you wanted.

You can check out the full example [here](https://gist.github.com/aivascu/1138df84bdf3a9ceb4134373021b14ec).

```cs
class FreezingNode : ISpecimenBuilderNode
{
    private readonly ISpecimenBuilder builder;
    private readonly IRequestSpecification specification;
    private CompositeSpecimenBuilder frozenBuilders = new CompositeSpecimenBuilder();

    public FreezingNode(ISpecimenBuilder builder, IRequestSpecification specification)
    {
        this.builder = builder ?? throw new ArgumentNullException(nameof(builder));
        this.specification = specification ?? throw new ArgumentNullException(nameof(specification));
    }

    public ISpecimenBuilderNode Compose(IEnumerable<ISpecimenBuilder> builders)
    {
        var composedBuilder = new CompositeSpecimenBuilder(builders);
        return new FreezingNode(composedBuilder, specification);
    }

    public object Create(object request, ISpecimenContext context)
    {
        if (!specification.IsSatisfiedBy(request))
        {
            return builder.Create(request, context);
        }
        else
        {
            object result = frozenBuilders.Create(request, null);
            if (result is not NoSpecimen)
            {
                return result;
            }
            result = builder.Create(request, context);
            frozenBuilders = new CompositeSpecimenBuilder(
                frozenBuilders,
                new FilteringSpecimenBuilder(
                    new FixedBuilder(result),
                    new EqualRequestSpecification(request)));
            return result;
        }
    }

    public IEnumerator<ISpecimenBuilder> GetEnumerator()
    {
        yield return builder;
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }
}
```

**Reply by dmtrrk** [reply]
this solution looks perfect. thank you for your help and time.
Now, by default I can scan the constructor of my `sut`, and freeze all that types
